### ISSUE TITLE
How much do you value the ability to block freely on the compute pool?

### ISSUE THREAD

**Issue by vasilmkd** [typelevel/cats-effect#1499]
This is a slightly long read, but please bear with me, I believe it's a worthwhile topic to be discussed. Thank you in advance.

As an exercise, please replace the following code in `IORuntimeCompanionPlatform` on the JVM.

```scala
  def createDefaultComputeThreadPool(
      self: => IORuntime,
      threadPrefix: String = "io-compute"): (ExecutionContext, () => Unit) = {
    val threads = 1
    val threadCount = new AtomicInteger(0)
    val executor = Executors.newFixedThreadPool(threads, { (r: Runnable) =>
      val t = new Thread(r)
      t.setName(s"${threadPrefix}-${threadCount.getAndIncrement()}")
      t.setDaemon(true)
      t
    })
    (ExecutionContext.fromExecutorService(executor), { () => executor.shutdown() })
  }
```

What this does, is it makes the default compute pool for the IO runtime to be a single threaded fixed size pool (think Javascript runtime).

If you now run `coreJVM/test` with this configuration, you will see that `DispatcherSpec` now fails.

This happens because of the following implicit assumption by `Dispatcher`: the runtime is able to guarantee that any `F[A]` which is to be executed by the Dispatcher can be `start`ed and _definitely transferred to a different carrier thread_ for execution. That thread will execute the effect and in turn notify the blocked compute thread that the effect has been completed. In this case, obviously, there is no other thread to pick up the work, resulting in a deadlock. But even then, what if the other threads, in a very unlucky situation, are all executing a `Dispatcher` action, or any action similar in semantics to `unsafeRunSync` (block a thread and wait for notification)? It will again lead to a deadlock, no matter the number of threads.

Why is this relevant to me? We have artificially set the minimum number of threads in the compute pool to be 2, in order to combat slow single threaded CI runtimes. This was way back before the `Dispatcher` was developed, which has made us implicitly depend on the fact that there will always be a thread to finish the work for the dispatched effect. I believe this fact should not be assumed of the underlying runtime, but I'd like to ask you another question.

Let's say a library client intentionally creates this situation, a single threaded runtime and decides to use `Dispatcher`. At that point, how do you explain the mess to the user? What if the user writes a similar type of code, just not using `Dispatcher` (again, block a thread and wait for a notification from some other thread, but that thread doesn't exist, so you end up waiting forever)? Who is at fault, is it the user, is it the runtime? I'd really like to know your thoughts. Thank you.

Finally, I'll try to quickly explain my view on this, as well as my motivation. I consider blocking on the compute pool to be evil, however, I also understand that we are running on the JVM, where this is extremely common behavior. Not everything can be annotated with `F.blocking` and I understand that (even though that doesn't exactly help in this situation, becausе `blocking` can only be evaluated if the spawned fiber ever starts execution, which doesn't ever happen in the situation I explained).

However, blocking on the compute pool prevents us from doing some seriously impressive performance optimizations for the IO runtime. I have found a way to reduce contention in the pool much further than the work stealing pool ever could. This results in ~40% (and this was an extremely WIP benchmark, I have done even more optimizations since) throughput improvements compared to the _work stealing pool_ on a 16 core ARM machine (largest machine I could test this on), which I strongly believe is not negligible and we should take advantage of it. This pool is different from the work stealing pool in the sense that it cannot guarantee this fact that another carrier thread might pick up the fiber that executes the dispatched effect. On the work stealing pool, that fiber could be stolen by another carrier thread. However, even this does not guarantee progress in the case that _all_ carrier threads are executing an `IO#unsafeRunSync` type of action (edit: see next comment).

If you are curious how far we've come, on that same ARM machine, the work stealing thread pool is 30x faster than a fixed thread pool executor with the same number of threads, while the new runtime I'm developing is 42x faster.

Thank you for your patience and attention, if you have made it this far. Curious to hear your thoughts.

**Comment by vasilmkd** [comment]
In the meantime, I have engineered the worst case scenario I described in my post. We have it in the tests even.

```scala
"run multiple IOs in parallel" in real {
  val num = 100

  for {
    latches <- (0 until num).toList.traverse(_ => Deferred[IO, Unit])
    awaitAll = latches.parTraverse_(_.get)

    // engineer a deadlock: all subjects must be run in parallel or this will hang
    // Vasil's slightly snarky comment: just not too parallel
    subjects = latches.map(latch => latch.complete(()) >> awaitAll)

    _ <- {
      val rec = Dispatcher[IO] flatMap { runner =>
        Resource.liftF(subjects.parTraverse_(act => IO(runner.unsafeRunSync(act))))
      }

      rec.use(_ => IO.unit)
    }
  } yield ok
}
```

This is current behavior on `series/3.x` as of today.

This hangs on my machine. Who is at fault here?

**Comment by keynmol** [comment]
Gonna reference a conversation that just happened on Gitter:

https://gitter.im/typelevel/cats-effect?at=5fd90c9e6c2aad2754074f74

@vasilmkd I unintentionally engineered your worst case scenario in unrelated CE2 -> CE3 conversion.

I was using CE3 to simulate parallelism to test a library's ability to resolve a deadlock-inducing situation. (code under test uses `Await.result`)

Under CE2, running unbounded `parTraverse` of tasks was okay, under CE3 it was deadlocking. 

Changing all of the calls inside `parTraverse` to `IO.blocking` (instead of `IO.apply`) eventually resolved it, but it was a gotcha that seriously made me doubt the testing framework, until I discovered this issue :)

**Comment by Daenyth** [comment]
It seems reasonable that `IO(someIO.unsafeRunSync())` risks a deadlock in principle. However..

In my experience with CE2, that shape is pretty likely to happen indirectly in a codebase that uses *some* cats-effect but doesn't use it across the entire stack.

Some interfaces might use IO, some might not, and if they call each other, it's easy to arrive in that shape accidentally.

Having that break would be a pretty bad experience for new users

**Comment by SystemFw** [comment]
I'm torn on this. My main worry is that users that are just trying it out cats-effect (which would often happen in a mixed codebase), would get a deadlock out of this, and never use cats-effect again (and probably bad press it with "tried it once, locked up my whole app in 5 minutes")

> the work stealing thread pool is 30x faster than a fixed thread pool executor with the same number of threads, while the new runtime I'm developing is 42x faster.

I'd say that 30x faster and completely safe is a better default than 42x faster but unsafe and requires discipline and a certain level of expertise, so even if do we go ahead, I would want the thread stealing to be on by default, and this new runtime to be an addition that you can swap in, if at all possible

**Comment by vasilmkd** [comment]
Completely agree that the runtime should be as safe as possible. But not even work stealing or any fixed thread pool is safe for that matter. I hope that people are at least aware of that.

**Comment by keynmol** [comment]
I have a fair amount of experience with CE2 (as a user), and having a deadlock in CE3 made me gasp :) Which is not a scientific measure by all means, but I'm spending a considerable amount of time to understand why it is happening. 

And tbh I'm only doing this because I'm helping maintain a testing framework for CE2/CE3 so distinguishing between user/library/framework errors is important for us.

**Comment by vasilmkd** [comment]
@keynmol I'm working on resolving that specific deadlock at the moment.
